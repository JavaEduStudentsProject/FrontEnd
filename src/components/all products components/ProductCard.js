import React from 'react';
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button';
import ProductList from "../../services/ProductList";
// import "bootstrap/dist/css/bootstrap.min.css";
import {Link} from 'react-router-dom'


const ProductCard = (props) => {
    const flatProduct = ProductList.flatProduct(props.item);
    console.log(flatProduct)
    // const keys = Object.keys(flatProduct).filter(key => key !== "image" && key !== "description" && key !== "title"
    // && key !== "0" && key !== "1" && key !== "2" && key !== "3" && key !== "4");

    const keys = Object.keys(flatProduct).filter(key => key === "brand" || key === "description" || key === "category");

    const descriptionList = keys.map((key, index) =>
        keys[index] === "description" ?
            flatProduct[key].length < 90 ?
            <li key={flatProduct.id + index}>{flatProduct[key]}</li> :
            <li key={flatProduct.id + index}>{flatProduct[key].substring(0, 90)}...</li> :
            <li key={flatProduct.id + index}>{flatProduct[key]}</li>
    );

    function getImage() {
        return flatProduct.image[0] !== "h" ? require(`../../images/${flatProduct.image}`) : process.env.PUBLIC_URL + flatProduct.image;
    }

    let image = getImage();


    return (
        <>
            <Card className='card' border={"success"}>
                <Card.Img className='product-img' variant="top" alt={flatProduct.title} src={image}/>
                <Card.Title className="product-name">{flatProduct.title}</Card.Title>
                <Card.Body className='product-description'>
                    {/*<Card.Text>*/}
                    <ul>
                        {descriptionList}
                    </ul>
                    {/*</Card.Text>*/}
                </Card.Body>
                <Card.Subtitle className="product-price">{flatProduct.price} $</Card.Subtitle>

                <Button className="product-button" variant="primary">
                    <Link className="link-button" to={`/product/${flatProduct.id}`}>Просмотр</Link>
                </Button>


                <Button onClick={() => props.incrementProductCount(flatProduct.id)} className="product-button"
                        variant="primary">
                    В корзину
                </Button>

            </Card>
        </>
    );
};

export default ProductCard;