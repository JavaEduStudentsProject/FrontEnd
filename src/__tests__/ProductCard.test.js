import {render, screen} from '@testing-library/react';

import ProductCard from '../components/all products components/ProductCard'
import {MemoryRouter} from "react-router";
import '@testing-library/jest-dom/extend-expect';

const data = {
        "id": 4,
        "title": "Kids Casual Slim Fit",
        "category": "clothing",
        "price": "15.99",
        "filter_features":
            {
                "subCategory": "kid's clothing",
                "material": "cotton",
                "size": "28"
            },
        "non-filter_features": {
            "rating": {
                "rate": "2.1",
                "count": "430"
            },
            "color": "red",
            "country": "Bangladesh",
            "style": "casual"
        },
        "description": "The color could be slightly different between on the screen and in practice. / Please note that body builds vary by person, therefore, detailed size information should be reviewed below on the product description.",
        "image": "https://fakestoreapi.com/img/71YXzeOuslL._AC_UY879_.jpg"
    }

    const emptyData = {}

describe('ProductCard component', () => {
    test('ProductCard renders', () => {
        render(<ProductCard item={data}/>, {wrapper: MemoryRouter});

        const linkElement = screen.getByText(/Kids Casual Slim Fit/i);

        // eslint-disable-next-line jest/valid-expect
        expect(linkElement).toBeInTheDocument();

        expect(screen.getByRole('list')).toBeInTheDocument();
    })

    test('List Product card renders without data', () => {
        render(<ProductCard />, {wrapper: MemoryRouter});

        expect(screen.queryByRole('list')).toBeNull();
    })

    test('List snapshot', () => {
        const view = render(<ProductCard item={data}/>, {wrapper: MemoryRouter});
        expect(view).toMatchSnapshot();
    })
})