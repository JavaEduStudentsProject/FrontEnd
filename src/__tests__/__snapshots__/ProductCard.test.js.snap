// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard component List snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card card border-success"
      >
        <img
          alt="Kids Casual Slim Fit"
          class="card-img-top product-img"
          src="https://fakestoreapi.com/img/71YXzeOuslL._AC_UY879_.jpg"
        />
        <div
          class="product-name card-title h5"
        >
          Kids Casual Slim Fit
        </div>
        <div
          class="product-description card-body"
        >
          <ul>
            <li>
              id
              : 
              4
            </li>
            <li>
              category
              : 
              clothing
            </li>
            <li>
              price
              : 
              15.99
            </li>
            <li>
              subCategory
              : 
              kid's clothing
            </li>
            <li>
              material
              : 
              cotton
            </li>
            <li>
              size
              : 
              28
            </li>
            <li>
              rate
              : 
              2.1
            </li>
            <li>
              count
              : 
              430
            </li>
            <li>
              color
              : 
              red
            </li>
            <li>
              country
              : 
              Bangladesh
            </li>
            <li>
              style
              : 
              casual
            </li>
          </ul>
        </div>
        <div
          class="product-price card-subtitle h6"
        >
          15.99
           $
        </div>
        <button
          class="product-button btn btn-primary"
          type="button"
        >
          <a
            class="link-button"
            href="/product/4"
          >
            Просмотр
          </a>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card card border-success"
    >
      <img
        alt="Kids Casual Slim Fit"
        class="card-img-top product-img"
        src="https://fakestoreapi.com/img/71YXzeOuslL._AC_UY879_.jpg"
      />
      <div
        class="product-name card-title h5"
      >
        Kids Casual Slim Fit
      </div>
      <div
        class="product-description card-body"
      >
        <ul>
          <li>
            id
            : 
            4
          </li>
          <li>
            category
            : 
            clothing
          </li>
          <li>
            price
            : 
            15.99
          </li>
          <li>
            subCategory
            : 
            kid's clothing
          </li>
          <li>
            material
            : 
            cotton
          </li>
          <li>
            size
            : 
            28
          </li>
          <li>
            rate
            : 
            2.1
          </li>
          <li>
            count
            : 
            430
          </li>
          <li>
            color
            : 
            red
          </li>
          <li>
            country
            : 
            Bangladesh
          </li>
          <li>
            style
            : 
            casual
          </li>
        </ul>
      </div>
      <div
        class="product-price card-subtitle h6"
      >
        15.99
         $
      </div>
      <button
        class="product-button btn btn-primary"
        type="button"
      >
        <a
          class="link-button"
          href="/product/4"
        >
          Просмотр
        </a>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
